cylc/flow/task_pool.py

```diff
@@ -1477,304 +1450,530 @@ def can_spawn(self, name: str, point: 'PointBase') -> bool:
+        if not self.config.get_taskdef(name).is_valid_point(point):
+            LOG.warning(
+                self.ERR_PREFIX_TASK_NOT_ON_SEQUENCE.format(
+                    name, point
+                )
+            )
+            return False
+
         return True
 
+    def _get_task_history(self, name, point, flow_nums):
+        """Get details of previous submits for this task.
+
+        """
+        info = self.workflow_db_mgr.pri_dao.select_prev_instances(
+            name, str(point)
+        )
+        try:
+            submit_num = max(s[0] for s in info)
+        except ValueError:
+            # never spawned before in any flow
+            submit_num = 0
+
+        prev_completed = False  # did not complete in the flow
+        prev_flow_wait = False  # did not wait in the flow
+        # orig_fnums = set()
+
+        for snum, f_wait, old_fnums, is_complete in info:
+            if set.intersection(flow_nums, old_fnums):
+                # matching flows
+                prev_completed = is_complete
+                prev_flow_wait = f_wait
+                # orig_fnums = old_fnums
+                if not prev_completed:
+                    # There may be multiple entries with flow overlap due
+                    # to merges (they'll have have same snum and f_wait);
+                    # keep going to find the complete one, if any .
+                    continue
+                LOG.warning(
+                    f"{point}/{name} already completed for flow"
+                    f" {stringify_flow_nums(flow_nums, full=True)} via"
+                    f" {point}/{name}/{snum:02d}"
+                    f"{stringify_flow_nums(old_fnums, full=True)}"
+                )
+                break
+
+        return submit_num, prev_completed, prev_flow_wait
+
+    def _load_historical_outputs(self, itask):
+        """Load a task's historical outputs from the DB."""
+        info = self.workflow_db_mgr.pri_dao.select_task_outputs(
+            itask.tdef.name, str(itask.point))
+        if not info:
+            # task never ran before
+            self.db_add_new_flow_rows(itask)
+        else:
+            for outputs_str, fnums in info.items():
+                if itask.flow_nums.intersection(fnums):
+                    for msg in json.loads(outputs_str):
+                        itask.state.outputs.set_completed_by_msg(msg)
+
     def spawn_task(
         self,
         name: str,
         point: 'PointBase',
         flow_nums: Set[int],
         force: bool = False,
-        is_manual_submit: bool = False,
         flow_wait: bool = False,
     ) -> Optional[TaskProxy]:
-        """Spawn point/name. Return the spawned task, or None.
+        """Spawn a task if not already completed for this flow, or if forced.
+
+        The creates the task proxy but does not add it to the pool.
+
+        If completed previously with flow wait, just try to spawn children.
+
+        # (TODO - reuse "cylc set" machinery for this?)
```

Comment:
TODO.

Can stick an issue up for this if it's ok for now.

Reply:
#5971